MotherOfAllQueries.sql

```diff
@@ -0,0 +1,407 @@
+FROM branches
+	JOIN banks on branches.BankId = banks.Id		
+WHERE branches.CityId = 2
+GO
+
+-- Task 3
+SELECT clients.Name AS ClientName, cards.Balance, banks.Name AS BankName
+FROM cards
+	JOIN accounts on cards.AccountId = accounts.Id
+	JOIN clients on accounts.ClientId = clients.Id
+	JOIN banks on accounts.BankId = banks.Id
+GO
+
+-- Task 4
+SELECT accounts.Id, accounts.Balance AS AccountBalance, COALESCE(SUM(cards.Balance),0) AS CardsBalance, accounts.Balance - COALESCE(SUM(cards.Balance),0) AS Mismatch
+FROM accounts
+	LEFT JOIN cards on accounts.Id = cards.AccountId
+GROUP BY accounts.Id, accounts.Balance
+HAVING accounts.Balance - COALESCE(SUM(cards.Balance),0) != 0
+GO
+
+-- Task 5 (using GROUP BY)
+SELECT socialGroups.Id, socialGroups.Name, COUNT(cards.Id) AS CardsCount
+FROM socialGroups
+	LEFT JOIN clients on clients.SocialGroupId = socialGroups.Id
+	LEFT JOIN accounts on accounts.ClientId = clients.Id
+	LEFT JOIN cards on cards.AccountId = accounts.Id
+GROUP BY socialGroups.Id, socialGroups.Name
+GO
+
+-- Task 5 (using subquery)
+SELECT socGroupsHigh.Id, socGroupsHigh.Name, (
+	SELECT COUNT(cards.Id) 
+	FROM socialGroups
+		LEFT JOIN clients on clients.SocialGroupId = socialGroups.Id
+		LEFT JOIN accounts on accounts.ClientId = clients.Id
+		LEFT JOIN cards on cards.AccountId = accounts.Id
+	WHERE socialGroups.Id = socGroupsHigh.Id) AS CardsCount
+FROM socialGroups as socGroupsHigh
+GO
+
+-- Task 6 
+CREATE PROCEDURE Add10$ToSocGroup 
+	@SocGroupId INT
+AS
+BEGIN
+	IF @SocGroupId < 1
+		BEGIN
+			PRINT 'Invalid social group ID'
+		END;
+	ELSE
+		IF NOT EXISTS (SELECT * 
+			FROM socialGroups
+			WHERE socialGroups.Id = @SocGroupId)
+			BEGIN
+				PRINT 'Social group with this ID does not exist'
+			END;
+		ELSE
+		IF NOT EXISTS (SELECT *
+			FROM clients
+			WHERE clients.SocialGroupId = @SocGroupId)
+			BEGIN
+				PRINT 'No clients of that social group found'
+			END;
+			ELSE
+				IF NOT EXISTS (SELECT *
+					FROM accounts
+						JOIN clients on accounts.ClientId = clients.Id
+					WHERE clients.SocialGroupId = @SocGroupId)
+					BEGIN
+						PRINT 'No accounts associated with clients of that social group found'
+					END
+				ELSE
+				BEGIN
+					UPDATE accounts
+					SET Balance += 10
+					FROM accounts
+						JOIN clients on clients.Id = accounts.ClientId
+					WHERE clients.SocialGroupId = @SocGroupId
+				END;
+END;
+GO
+
+SELECT accounts.Balance
+FROM accounts
+
+EXEC Add10$ToSocGroup 1
+
+SELECT accounts.Balance
+FROM accounts
+GO
+
+-- Task 7
+SELECT accounts.Id, clients.Name, clients.Surname, accounts.Balance - COALESCE(SUM(cards.Balance),0) AS AvailableFunds
+FROM accounts
+	JOIN clients on accounts.ClientId = clients.Id
+	LEFT JOIN cards on cards.AccountId = accounts.Id
+GROUP BY accounts.Id, clients.Name, clients.Surname, accounts.Balance
+GO
+
+-- Task 8
+CREATE PROCEDURE TransferMoneyToCardFromAcc 
+	@AccountId INT,
+	@CardId INT,
+	@Amount MONEY
+AS
+BEGIN
+	BEGIN TRY
+	BEGIN TRANSACTION
```

Comment:
it's better to start the transaction right before the modifying data, just validation can be outside of it